contracts/token/ERC721/ERC721.sol

```diff
@@ -330,32 +296,53 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata {
      *
      * Emits a {Transfer} event.
      */
-    function _transfer(address from, address to, uint256 tokenId) internal virtual {
+    function _transfer(address from, address to, uint256 tokenId) internal {
         require(ERC721.ownerOf(tokenId) == from, "ERC721: transfer from incorrect owner");
         require(to != address(0), "ERC721: transfer to the zero address");
 
-        _beforeTokenTransfer(from, to, tokenId, 1);
+        _update(from, to, tokenId, 1);
+    }
 
-        // Check that tokenId was not transferred by `_beforeTokenTransfer` hook
-        require(ERC721.ownerOf(tokenId) == from, "ERC721: transfer from incorrect owner");
+    /**
+     * @dev Transfers `tokenId` from `from` to `to`, or alternatively mints (or burns) if `from` (or `to`) is
+     * the zero address. All customizations to transfers, mints, and burns should be done by overriding this function.
+     *
+     * Emits a {Transfer} event.
+     */
+    function _update(address from, address to, uint256 tokenId, uint256 batchSize) internal virtual {
+        if (batchSize > 1) {
```

Comment:
I'm not a fan of this switch.

In all cases
```
if (from != address(0)) {
    _balances[from] -= batchSize;
}
if (to != address(0)) {
    _balances[to] += batchSize;
}
```
(the unchecked part might be challenging)

**But more importantly, there is no check that `batchSize > 0`!**

If batchSize is 0, then we go in the else section, which will burn the token (oups)